IdentityCore/src/pop_manager/MSIDDevicePopManager.m

```diff
@@ -0,0 +1,216 @@
+        NSData *hashedData = [jwkData msidSHA256];
+        _kid = [hashedData msidBase64UrlEncodedString];
+    }
+    
+    return _kid;
+}
+
+- (NSString *)createSignedAccessToken:(NSString *)accessToken
+                           httpMethod:(NSString *)httpMethod
+                           requestUrl:(NSString *)requestUrl
+                                nonce:(NSString *)nonce
+                                error:(NSError *__autoreleasing * _Nullable)error
+{
+    NSString *kid = self.kid;
+    
+    if (!kid)
+    {
+        [self logAndFillError:@"Failed to create signed access token, unable to generate kid." error:error];
+        return nil;
+    }
+    
+    NSURL *url = [NSURL URLWithString:requestUrl];
+    if (!url)
+    {
+        [self logAndFillError:[NSString stringWithFormat:@"Failed to create signed access token, invalid request url : %@.",requestUrl] error:error];
+        return nil;
+    }
+    
+    NSString *host = url.host;
+    if (!host)
+    {
+        [self logAndFillError:[NSString stringWithFormat:@"Failed to create signed access token, invalid request url : %@.",requestUrl] error:error];
+        return nil;
+    }
+    
+    NSString *path = url.path;
+    if (!path)
+    {
+        [self logAndFillError:[NSString stringWithFormat:@"Failed to create signed access token, invalid request url : %@.",requestUrl] error:error];
+        return nil;
+    }
+    
+    NSString *publicKeyModulus = [self.keyPair getKeyModulus:self.keyPair.publicKeyRef];
+    if (!publicKeyModulus)
+    {
+        [self logAndFillError:@"Failed to create signed access token, unable to read public key modulus." error:error];
+        return nil;
+    }
+    
+    NSString *publicKeyExponent = [self.keyPair getKeyExponent:self.keyPair.publicKeyRef];
+    if (!publicKeyExponent)
+    {
+        [self logAndFillError:@"Failed to create signed access token, unable to read public key exponent." error:error];
+        return nil;
+    }
+    
+    NSDictionary *header = @{
+                             @"alg" : @"RS256",
+                             @"typ" : @"JWT",
+                             @"kid" : kid
+                             };
+    
+    NSDictionary *payload = @{
+                              @"at" : accessToken,
+                              @"cnf": @{
+                                      @"jwk":@{
+                                          @"kty" : @"RSA",
+                                          @"n" : publicKeyModulus,
+                                          @"e" : publicKeyExponent
+                                      }
+                              },
+                              @"ts" : [NSString stringWithFormat:@"%lu", (long)[[NSDate date] timeIntervalSince1970]],
+                              @"m" : httpMethod,
+                              @"u" : host,
+                              @"p" : path,
+                              @"nonce" : nonce,
```

Comment:
this will crash if nonce == nil

Reply:
nonce is initialized in the constructor of pop manager so I think we are fine.